### ISSUE TITLE
Feature request: Custom URL / config or QR code to pre-configure the Traccar Client

### ISSUE THREAD

**Issue by ovizii** [traccar/traccar-client-android#355]
I originally raised the question on the forum: https://www.traccar.org/forums/topic/is-it-possible-to-deploy-a-mobile-client-pre-configured-or-send-configuration-file/ 

but was advised to open a feature request here.  What I am looking for is a way to pre-configure a mobile traccar client. There are several way but some ideas are: 

a) create user + device on traccar server. click button to email config to user. User opens config on mobile device. traccar client gets configured.

b) create user + device on traccar server. click button to send email with QR code to user. User scans QR code and 1) traccar client gets configured if installed or if not 2) traccar client gets installed and configured. 

As far as my experience goes, Duo Security uses b) to add their 2FA clients to mobile phones and it works quite efficiently...

**Comment by stefanb** [comment]
Or use a specially crafted url, eg:
https://www.traccar.org/config?url=https://customserver.example.org&clientid=123&freq=60
this can be read shared as QR codes or links within emails or offered within the user's web GUI.

When user visits the link on the device with client installed it can be opened with the client and then the client configures the parameters as instructed by the query string parameters. 

Other parameters can be added as desired to the query string.

If user opens the link in the browser he is given instructions to open the link within the app and how to set the configuration in the client (interpreting the query string parameters).

After configuring the client does not need to make any http request, but can report a success or make a quick test.

Here is how to create such an app link:
https://developer.android.com/studio/write/app-link-indexing

**Comment by reslin** [comment]
A special URL would be associated to a fixed App-Link. You would have to provide extra APKs for each configuration-URL.

**Comment by groupnet** [comment]
> A special URL would be associated to a fixed App-Link. You would have to provide extra APKs for each configuration-URL.

Not necessarily: I would stick to @stefanb suggestion with a specially crafted URL that can be also queried from a web browser. Like this the URL could not only be registered as intent (for Android), but also be interpreted by the web-server (in the example by a page at https://www.traccar.org/config) when opened in a web browser. This page could serve information on what to do with this config link, eg. that it is supposed to be opened with Traccar from a mobile phone. I guess this is a nice feature, especially if the link is shared via email (and opened on a desktop).

So opening the example link https://www.traccar.org/config?url=... in a browser could give information like:
```
This is a configuration link for the Traccar app. Please install the app first:
[LINK_TO_GOOGLE_PLAY]
[LINK_TO_APPLE_ITUNES]
```

Additionally it could even render the GET parameters into a mock-up view of the Traccar settings page to visually guide the user through the config:
```
If you have Traccar already installed on your phone,
please open the settings enter the following parameters:

      URL:  https://customserver.example.org
Client Id:  123
Frequency:  60
```

By the way: the GET parameters would need to be encoded, eg. with URLEncoder (https://stackoverflow.com/a/3286128/5157195).

**Comment by stefanb** [comment]
I noticed there are quite many unrelated changes in https://github.com/reslin/traccar-client-android/commit/f22fc2369a2454e7ba5dc013ed00b06262b9e27e , possibly due to different editor that is normally used before, making it harder for review and more importantly: prone for conflicts with the master branch. I suggest you create a WIP (Work In Progress) pull request from that branch.

For the out-of-the-box initial configuration perhaps we could use play store parameters, just like the referrer tracking, outlined at
 https://developers.google.com/analytics/devguides/collection/android/v4/campaigns

If there is no other way to pass some custom parameters during the app installation we could abuse the above referrer parameters to include the customized configuration settings ;-)

**Comment by reslin** [comment]
> I noticed there are quite many unrelated changes in [reslin@f22fc23](https://github.com/reslin/traccar-client-android/commit/f22fc2369a2454e7ba5dc013ed00b06262b9e27e) , possibly due to different editor that is normally used before, making it harder for review and more importantly: prone for conflicts with the master branch. I suggest you create a WIP (Work In Progress) pull request from that branch.
> 

Yes, that's odd. I'm using Android Studio 3.2.1, but since this is actually my very first contribution I may have overseen some options for formatting (resp. _not_ formatting) the source files before commit. I'm sorry for that.

> For the out-of-the-box initial configuration perhaps we could use play store parameters, just like the referrer tracking, outlined at
> https://developers.google.com/analytics/devguides/collection/android/v4/campaigns
> 
> If there is no other way to pass some custom parameters during the app installation we could abuse the above referrer parameters to include the customized configuration settings ;-)

I'll have to read through this thoroughly because I don't get the concept right now. Do you think it's worth the effort? Or is the topic really more like "configure once and forget about it" ...

**Comment by reslin** [comment]
I'll roll out another branch and watch for unnecessary alterations of the master sources. (When back home in the evening.)

**Comment by reslin** [comment]
Ok, there's a new branch "remote-conf" in which I tried not to make any deletions of original master-source. I also refrained from doing "too much", especially altering the displaying of preferences etc.

Should be neat and clean now: [reslin@7c10f85](https://github.com/traccar/traccar-client-android/commit/7c10f85ecd3dbca6ac73721021f8e56f8460f96b)

**Comment by andsdoug** [comment]
Another reason to implement configuration via QR code is the convenience for users who need to set up multiple devices with the same settings. With the option to generate a QR code with the desired configurations, users can simply scan the code with their devices and all parameters will be automatically configured without the need to manually enter information on each individual device. This saves time and effort for users and reduces the likelihood of configuration errors. Furthermore, this functionality has the potential to attract more users to Traccar due to its user-friendly setup process.

**Comment by redge76** [comment]
I'm missing this feature too. It pretty time consuming to help people to configure their phone with the correct urls & settings.
A QR code with the correct configuration url would be a good solution.
You can encode the parameters in the url or better just send the user to a page on the traccar server.
This page would send back the correct settings.

**Comment by pissten** [comment]
this feature would be great! Would make the client and sever so much more userfriendly to set-up. In terms of automation, the client ids and server url could be autogenerated from a backend tool, and sent to the clients as a QR-code, just love the idea!
Are there any working being put into this lately? or is it stall as of 2023?

**Comment by redge76** [comment]
https://github.com/traccar/traccar-client-android/pull/486

**Comment by twproject** [comment]
Hi all,
I didn't understand if there is a way to send an email to a remote user with a traccar url to auto configure the app.
Our needs is to send an email to every user with a link to configure server site, client Id (it will be the race Id) and the frequency )
We saw the qr code but it's not easy to read a qr code from the same app device

Thanks

**Comment by tananaev** [comment]
Not yet.
